File: frontend/lib/api/linkedin.ts
// Agente LinkedIn (P5 §6.C) — admin no studio: CRUD + publicar. NÃO auto-posta
// no LinkedIn; gera rascunhos pra Página da Reative e pro perfil pessoal.

import { request } from './client';
import type {
  GerarImagemLinkedinRequest,
  LinkedinBriefRequest,
  LinkedinConta,
  LinkedinGerarRequest,
  LinkedinPost,
  LinkedinPostCreate,
  LinkedinPostUpdate,
  LinkedinRedacao,
  LinkedinStatus,
} from '../types';

const BASE = '/api/linkedin';
const T = 12_000;

export const linkedinApi = {
  linkedinListar(
    status?: LinkedinStatus,
    conta?: LinkedinConta,
  ): Promise<LinkedinPost[]> {
    const qs = new URLSearchParams();
    if (status) qs.set('status', status);
    if (conta) qs.set('conta', conta);
    const sufixo = qs.toString() ? `?${qs.toString()}` : '';
    return request<LinkedinPost[]>(`${BASE}/posts${sufixo}`, { timeoutMs: T });
  },

  linkedinGet(id: string): Promise<LinkedinPost> {
    return request<LinkedinPost>(`${BASE}/posts/${id}`, { timeoutMs: T });
  },

  linkedinCriar(body: LinkedinPostCreate): Promise<LinkedinPost> {
    return request<LinkedinPost>(`${BASE}/posts`, {
      method: 'POST',
      body,
      timeoutMs: T,
    });
  },

  linkedinAtualizar(id: string, body: LinkedinPostUpdate): Promise<LinkedinPost> {
    return request<LinkedinPost>(`${BASE}/posts/${id}`, {
      method: 'PUT',
      body,
      timeoutMs: T,
    });
  },

  linkedinMudarStatus(id: string, status: LinkedinStatus): Promise<LinkedinPost> {
    return request<LinkedinPost>(`${BASE}/posts/${id}/status`, {
      method: 'PATCH',
      body: { status },
      timeoutMs: T,
    });
  },

  linkedinRemover(id: string): Promise<void> {
    return request<void>(`${BASE}/posts/${id}`, {
      method: 'DELETE',
      timeoutMs: T,
    });
  },

  // ── Agente (L1) ─────────────────────────────────────────────────
  linkedinRedigir(brief: LinkedinBriefRequest): Promise<LinkedinRedacao> {
    return request<LinkedinRedacao>(`${BASE}/redigir`, {
      method: 'POST',
      body: brief,
      timeoutMs: 90_000, // geração pela LLM é lenta
    });
  },

  // ── Coordenador (L2) ────────────────────────────────────────────
  linkedinGerar(req: LinkedinGerarRequest): Promise<LinkedinPost[]> {
    return request<LinkedinPost[]>(`${BASE}/gerar`, {
      method: 'POST',
      body: req,
      timeoutMs: 180_000, // gera vários posts em sequência
    });
  },

  // ── Direção de arte (L5) ────────────────────────────────────────
  linkedinSugerirMidia(id: string): Promise<LinkedinPost> {
    return request<LinkedinPost>(`${BASE}/posts/${id}/midia/sugerir`, {
      method: 'POST',
      timeoutMs: 90_000,
    });
  },

  linkedinGerarImagem(
    id: string,
    req: GerarImagemLinkedinRequest,
  ): Promise<LinkedinPost> {
    return request<LinkedinPost>(`${BASE}/posts/${id}/imagem`, {
      method: 'POST',
      body: req,
      timeoutMs: 120_000,
    });
  },
};

File: backend/app/api/routers/linkedin.py
"""API do agente LinkedIn (P5 §6.C) — CRUD/publicar pelo studio.

Autenticada (`linkedin.editar`). É a mesa de edição do Pablo: criar/editar
rascunho, marcar como publicado (depois de postar na mão), arquivar. O agente de
IA (L1+) escreve via estes mesmos services. NÃO auto-posta no LinkedIn.
"""
from fastapi import APIRouter, Depends, HTTPException, Query

from app.api.dependencies.auth import require_permission
from app.api.schemas.linkedin import (
    GerarImagemLinkedinRequest,
    LinkedinBriefRequest,
    LinkedinGerarRequest,
    LinkedinPostCreate,
    LinkedinPostOut,
    LinkedinPostUpdate,
    LinkedinRedacao,
    LinkedinStatusUpdate,
)
from app.api.services import linkedin_service
from app.api.services.linkedin_service import LinkedinError

router = APIRouter(
    prefix="/api/linkedin",
    tags=["linkedin"],
    dependencies=[Depends(require_permission("linkedin.editar"))],
)


def _handle(e: Exception) -> HTTPException:
    if isinstance(e, LinkedinError):
        msg = str(e)
        status = 404 if "não encontrad" in msg.lower() else 400
        return HTTPException(status_code=status, detail=msg)
    return HTTPException(status_code=500, detail=f"{type(e).__name__}: {e}")


@router.get("/posts", response_model=list[LinkedinPostOut], summary="Lista posts (filtra por status/conta)")
async def listar(
    status: str | None = Query(None),
    conta: str | None = Query(None),
) -> list[LinkedinPostOut]:
    try:
        return await linkedin_service.admin.listar(status=status, conta=conta)
    except Exception as e:
        raise _handle(e)


@router.get("/posts/{post_id}", response_model=LinkedinPostOut, summary="Detalhe do post")
async def get(post_id: str) -> LinkedinPostOut:
    try:
        return await linkedin_service.admin.get(post_id)
    except Exception as e:
        raise _handle(e)


@router.post("/redigir", response_model=LinkedinRedacao, summary="IA: brief → post (rascunho pro drawer)")
async def redigir(brief: LinkedinBriefRequest) -> LinkedinRedacao:
    try:
        return await linkedin_service.agente.redigir(brief)
    except Exception as e:
        raise _handle(e)


@router.post("/gerar", response_model=list[LinkedinPostOut], summary="IA autônoma: gera rascunhos (projetos/tendências)")
async def gerar(req: LinkedinGerarRequest) -> list[LinkedinPostOut]:
    try:
        return await linkedin_service.coordenador.gerar(req)
    except Exception as e:
        raise _handle(e)


@router.post("/posts", response_model=LinkedinPostOut, status_code=201, summary="Cria post (rascunho)")
async def criar(payload: LinkedinPostCreate) -> LinkedinPostOut:
    try:
        return await linkedin_service.admin.criar(payload)
    except Exception as e:
        raise _handle(e)


@router.put("/posts/{post_id}", response_model=LinkedinPostOut, summary="Edita post (patch parcial)")
async def atualizar(post_id: str, payload: LinkedinPostUpdate) -> LinkedinPostOut:
    try:
        return await linkedin_service.admin.atualizar(post_id, payload)
    except Exception as e:
        raise _handle(e)


@router.post("/posts/{post_id}/midia/sugerir", response_model=LinkedinPostOut, summary="IA: sugere a mídia ideal (direção de arte)")
async def sugerir_midia(post_id: str) -> LinkedinPostOut:
    try:
        return await linkedin_service.midia.sugerir(post_id)
    except Exception as e:
        raise _handle(e)


@router.post("/posts/{post_id}/imagem", response_model=LinkedinPostOut, summary="IA: gera a imagem do post → MinIO")
async def gerar_imagem(post_id: str, req: GerarImagemLinkedinRequest) -> LinkedinPostOut:
    try:
        return await linkedin_service.midia.gerar_imagem(
            post_id, prompt=req.prompt, alt=req.alt, aspect_ratio=req.aspect_ratio
        )
    except Exception as e:
        raise _handle(e)


@router.patch("/posts/{post_id}/status", response_model=LinkedinPostOut, summary="Muda status (publicar/arquivar)")
async def mudar_status(post_id: str, payload: LinkedinStatusUpdate) -> LinkedinPostOut:
    try:
        return await linkedin_service.admin.mudar_status(post_id, payload.status)
    except Exception as e:
        raise _handle(e)


@router.delete("/posts/{post_id}", status_code=204, summary="Apaga post")
async def deletar(post_id: str) -> None:
    try:
        await linkedin_service.admin.deletar(post_id)
    except Exception as e:
        raise _handle(e)
